pkg/workflow/controller/handlers/workflowrun/handler.go

```diff
@@ -149,47 +183,88 @@ func (h *Handler) ObjectDeleted(obj interface{}) {
 	return
 }
 
-// validate workflow run
-func validate(wfr *v1alpha1.WorkflowRun) bool {
-	// check workflowRef can not be nil
-	if wfr.Spec.WorkflowRef == nil {
-		log.WithField("name", wfr.Name).Error("WorkflowRef is nil")
-		return false
+// sendNotifications send notifications for workflowruns when:
+// * its workflow has notification config
+// * finish time after workflow controller starts
+// * notification status of workflowrun is not nil
+// If the returned notification status is nil, it means that there is no need to send notification.
+func (h *Handler) sendNotifications(wfr *v1alpha1.WorkflowRun) (map[string]v1alpha1.NotificationStatus, error) {
+	// No need to send notifications for workflowruns finished before workflow controller starts.
+	if wfr.Spec.WorkflowRef == nil || wfr.Status.Notifications != nil ||
```

Comment:
`wfr.Spec.WorkflowRef == nil` is an abnormal condition, I think we should report error instead of keep silent